### PR TITLE
feat: [CDS-57669]: fixing cert auth type for AKS kubelogin auth plugin.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Cert Auth type using AKS cloud provider did not work with kubelogin as it did not support PEM secret format. Hence we required AzureCli to provide access token to make the the kube config with exec format work. Azure Cli supports secret in PEM format. In this PR we are creating a new working directory inside `./repository/k8s/${ACTIVITY_ID}` and saving it as environment variable `AZURE_CONFIG_DIR` to store the azure token and session information. While executing the kubeconfig it will refer to this environment variable to to fetch the access token and use it to authenticate to K8s cluster.

PR link: https://cd.pr.harness.io/cds-52514-2/ng/account/kmpySmUISimoRrJL6NL73w/cd/orgs/default/projects/defaultproject/pipelines/k8sHelm/pipeline-studio/?storeType=INLINE&stageId=stage&sectionId=EXECUTION

Here are the various scenarios tested for this PR

- K8s Deployment without kubelogin and azure-cli installed on delegate
Screenshot 2023-03-24 at 12 02 34 PM
Screenshot 2023-03-24 at 12 02 46 PM
Screenshot 2023-03-24 at 12 03 08 PM

- K8s Deployment with kubelogin installed but not azure-cli on delegate
Screenshot 2023-03-24 at 12 04 58 PM
Screenshot 2023-03-24 at 12 05 09 PM

- K8s Deployment with kubelogin and azure-cli installed on delegate
Screenshot 2023-03-24 at 12 06 14 PM
Screenshot 2023-03-24 at 12 06 46 PM
Screenshot 2023-03-24 at 12 07 05 PM
Screenshot 2023-03-24 at 12 07 27 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)

### CODE REVIEW

**Comment by GoelPratyush99** [comment]
trigger messagecheck

File: 930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/taskhelper/AzureAsyncTaskHelper.java [hunk 1]
```diff
@@ -575,7 +575,11 @@ private KubernetesConfig getKubernetesConfigK8sCluster(AzureConfig azureConfig,
       } else if (userConfig.getExec() != null) {
         String apiServerId = Exec.getValueFromArgsList(userConfig.getExec().getArgs(), KUBECFG_ARGS_SERVER_ID);
         azureKubeConfig.setAadToken(fetchAksAADToken(azureConfig, apiServerId));
-        userConfig.setExec(updateAzureKubeconfig(userConfig.getExec(), azureConfig, apiServerId));
+        Map<String, String> env = new HashMap<>();
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/taskhelper/AzureAsyncTaskHelper.java:578]: What is the point of this env variable if it's always empty?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Comment by GoelPratyush99** [comment]
trigger ut0

**Comment by GoelPratyush99** [comment]
trigger ut5

**Comment by GoelPratyush99** [comment]
trigger ut6

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 11 of 14 files at r1, 31 of 38 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java` line 66 at r3](https://reviewable.io/reviews/harness/harness-core/45460#-NRXPdz70DFjHUtnn3eo:-NRXPdz70DFjHUtnn3ep:bkx0iyd) ([raw file](https://github.com/harness/harness-core/blob/835cc89057ea629c8c3aff7f665e4dfeb7cb8fba/960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java#L66)):*
> ```Java
>     if (KubeConfigAuthPluginHelper.runCommand(azureCliClientVersionCommand, logCallback, env)
>         && StringUtils.isNotEmpty(workingDirectory)) {
>       env.put(AZURE_CONFIG_DIR, workingDirectory);
> ```

Can we have this working directory as Paths.get(workingDirectory, AZURE_LOGIN_CONFIG_DIR_PATH) where AZURE_LOGIN_CONFIG_DIR_PATH is something like ".azure"
___
*[`970-api-services-beans/src/main/java/io/harness/azure/model/AzureConstants.java` line 405 at r3](https://reviewable.io/reviews/harness/harness-core/45460#-NRXPUbm3poN3o9DFNtl:-NRXPUbm3poN3o9DFNtm:bsayopb) ([raw file](https://github.com/harness/harness-core/blob/835cc89057ea629c8c3aff7f665e4dfeb7cb8fba/970-api-services-beans/src/main/java/io/harness/azure/model/AzureConstants.java#L405)):*
> ```Java
>   String AZURE_CLI_CMD = "az";
>   String AZURE_CONFIG_DIR = "AZURE_CONFIG_DIR";
>   String WORKING_DIR_BASE = "./repository/k8s/";
> ```

what is constant about?

**Comment by GoelPratyush99** [comment]
trigger ut1

**Comment by GoelPratyush99** [comment]
trigger ut6

**Comment by GoelPratyush99** [comment]
trigger ut1

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java` line 68 at r5](https://reviewable.io/reviews/harness/harness-core/45460#-NRb8PbK5u201_z48rke:-NRb8PbK5u201_z48rkf:b-aqrisc) ([raw file](https://github.com/harness/harness-core/blob/cd7eba6494f51fdac19731ce4b5ca6650770714d/960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java#L68)):*
> ```Java
>     if (KubeConfigAuthPluginHelper.runCommand(azureCliClientVersionCommand, logCallback, env)
>         && StringUtils.isNotEmpty(workingDirectory)) {
>       env.put(AZURE_CONFIG_DIR, Paths.get(workingDirectory, AZURE_LOGIN_CONFIG_DIR_PATH).toString());
> ```

And can we make sure that this is an absolute path?
___
*[`970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java` line 45 at r5](https://reviewable.io/reviews/harness/harness-core/45460#-NRb83im8vfWpNICc3kn:-NRb83im8vfWpNICc3ko:bg4c4f2) ([raw file](https://github.com/harness/harness-core/blob/cd7eba6494f51fdac19731ce4b5ca6650770714d/970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java#L45)):*
> ```Java
>     try {
>       return executeShellCommand(command, logCallback, environment);
>     } catch (Exception e) {
> ```

This indeed can be an issue, we should have a different block for interrupted exception otherwise this may mess up with the delegate scheduling logic
you can do something like:
catch (InterruptedException e) {
  Thread.currentThread().interrupt()
}

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/taskhelper/AzureAsyncTaskHelper.java` line 578 at r1](https://reviewable.io/reviews/harness/harness-core/45460#-NR7_J48FF0fgrtNzIM0-r1-578:-NRbManc3MQD7EQvfboZ:ba9j5b0) ([raw file](https://github.com/harness/harness-core/blob/21543f2f817abb4967dd4f231625df3dd1ba6dd3/930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/taskhelper/AzureAsyncTaskHelper.java#L578)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

What is the point of this env variable if it's always empty?
</blockquote></details>

It will not be empty for cert auth type.
___
*[`960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java` line 66 at r3](https://reviewable.io/reviews/harness/harness-core/45460#-NRXPdz70DFjHUtnn3eo:-NRbMfhA24bvlhNz6LjV:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/835cc89057ea629c8c3aff7f665e4dfeb7cb8fba/960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java#L66)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Can we have this working directory as Paths.get(workingDirectory, AZURE_LOGIN_CONFIG_DIR_PATH) where AZURE_LOGIN_CONFIG_DIR_PATH is something like ".azure"
</blockquote></details>

Done
___
*[`970-api-services-beans/src/main/java/io/harness/azure/model/AzureConstants.java` line 405 at r3](https://reviewable.io/reviews/harness/harness-core/45460#-NRXPUbm3poN3o9DFNtl:-NRbMiDz4HdEwujiMorI:ba3dfob) ([raw file](https://github.com/harness/harness-core/blob/835cc89057ea629c8c3aff7f665e4dfeb7cb8fba/970-api-services-beans/src/main/java/io/harness/azure/model/AzureConstants.java#L405)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

what is constant about?
</blockquote></details>

Removed
___
*[`970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java` line 45 at r5](https://reviewable.io/reviews/harness/harness-core/45460#-NRb83im8vfWpNICc3kn:-NRbMkB-1OgyjpeQqRKa:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/cd7eba6494f51fdac19731ce4b5ca6650770714d/970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java#L45)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

This indeed can be an issue, we should have a different block for interrupted exception otherwise this may mess up with the delegate scheduling logic
you can do something like:
catch (InterruptedException e) {
  Thread.currentThread().interrupt()
}
</blockquote></details>

Done.

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java` line 68 at r5](https://reviewable.io/reviews/harness/harness-core/45460#-NRb8PbK5u201_z48rke:-NRbQ6-F73glJjXaWSGW:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/cd7eba6494f51fdac19731ce4b5ca6650770714d/960-api-services/src/main/java/io/harness/azurecli/AzureCliClient.java#L68)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

And can we make sure that this is an absolute path?
</blockquote></details>

Done.

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java` line 46 at r6](https://reviewable.io/reviews/harness/harness-core/45460#-NRbR762-02cuB51tvU-:-NRbR762-02cuB51tvU0:b-qzg45e) ([raw file](https://github.com/harness/harness-core/blob/c5a0fe9b27f96b81848ba499dcbc8ac9a2af160d/970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java#L46)):*
> ```Java
>       return executeShellCommand(command, logCallback, environment);
>     } catch (InterruptedException e) {
>       Thread.currentThread().interrupt();
> ```

missing return statement

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java` line 46 at r6](https://reviewable.io/reviews/harness/harness-core/45460#-NRbR762-02cuB51tvU-:-NRbj24z042MAdcfzB-1:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c5a0fe9b27f96b81848ba499dcbc8ac9a2af160d/970-api-services-beans/src/main/java/io/harness/k8s/model/kubeconfig/KubeConfigAuthPluginHelper.java#L46)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

missing return statement
</blockquote></details>

Done.

**Review by abosii-harness — DISMISSED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: 49 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by GoelPratyush99** [comment]
trigger ut5

**Comment by GoelPratyush99** [comment]
trigger ut9

**Comment by GoelPratyush99** [comment]
trigger ut2

**Review by abosii-harness — APPROVED** [review]
> Reviewed 14 of 14 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: :shipit: complete! ✓

**Review by abhinavsingh1196 — APPROVED** [review]
> Reviewed 7 of 14 files at r1, 19 of 38 files at r2, 6 of 7 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, 1 of 2 files at r6, 14 of 14 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45460)* status: :shipit: complete! ✓

**Pull request merged**